Andrei react annotator modified

andrei-msd opened the pull request.

andrei-msd committed:
Allow multiple images to be uploaded onto annotator

andrei-msd committed:
Allow multiple images to be uploaded onto annotator

andrei-msd committed:
Allow multiple images to be uploaded onto annotator

andrei-msd committed:
Allow multiple images to be uploaded onto annotator

andrei-msd committed:
allow to add images to the annotator after initial upload

andrei-msd committed:
allow to add images to the annotator after initial upload

andrei-msd committed:
allow to add images to the annotator after initial upload

andrei-msd committed:
Fix adding string annotations to regions

andrei-msd committed:
Fix adding string annotations to regions

andrei-msd committed:
Fix adding string annotations to regions

andrei-msd committed:
Fix adding string annotations to regions

andrei-msd committed:
import from local and modify annotator functionalities

andrei-msd committed:
import from local and modify annotator functionalities

andrei-msd committed:
import from local and modify annotator functionalities

andrei-msd committed:
import from local and modify annotator functionalities

Commit:
rebase

Commit:
rebase

Commit:
Merge branch 'andrei-react-annotatote-modified' of https://github.com/michaeliwaikato/ufdl-frontend into andrei-react-annotatote-modified

andrei-msd committed:
Fix formatting and add documentation

andrei-msd committed:
Fix formatting and add documentation

andrei-msd committed:
Remove library from src

andrei-msd committed:
Remove library from src

andrei-msd committed:
Merging

andrei-msd committed:
Rename annotator component

andrei-msd committed:
Remove duplicate

andrei-msd committed:
Remove initiate state

andrei-msd committed:
Remove initial state

andrei-msd committed:
Move save function to add image button

yaserazfar reviewed (approved):
LGTM. Have tested the functionality with your branch in ufdl-image-annotator. Everything seems to be working.

yaserazfar commented:
Don't forget to squash before you merge please hahah

michaeliwaikato commented on frontend-react/src/components/annotate.js:
```diff
@@ -3,73 +3,164 @@ import ReactImageAnnotate from "react-image-annotate";
-    //       Uploading new images currently recreates the annotator and the annotations are cleared
-    state = {
+  state = {
+        uploaded: false,
```
This has inconsistent spacing in it (2 vs 4 spaces).

It should probably also be put in a constructor.

michaeliwaikato commented on frontend-react/src/components/annotate.js:
```diff
@@ -3,73 +3,164 @@ import ReactImageAnnotate from "react-image-annotate";
+  delete(annotateData){
+    //Set index to current displayed image
+    this.setState({
+      index: this.state.index = annotateData.selectedImage
```
It would be good if you could comment this to indicate this is intended to fix an error when the '=' isn't used. (Possibly state the error)

Edit: Probably should take a look at [this](https://reactjs.org/docs/faq-state.html) as it likely fixes some problems
Edits 2: This is definitely a hack and can easily be fixed with a redesign of the logic

michaeliwaikato commented on frontend-react/src/components/annotate.js:
```diff
@@ -3,73 +3,164 @@ import ReactImageAnnotate from "react-image-annotate";
+
+      //Remove current image from data images and decrement index
+      result.splice(annotateData.selectedImage,1);
+      this.setState({
+        data: result,
+        index: this.state.index = this.state.index - 1
```
This could be causing problems as indicated above.

michaeliwaikato commented on frontend-react/src/components/annotate.js:
```diff
@@ -3,73 +3,164 @@ import ReactImageAnnotate from "react-image-annotate";
+      });
+
+      //If there are no more images left in Annotator
+      if (this.state.index < 0){
+        //Set index back to 0
+        this.setState({
+          index: this.state.index = 0
```
Same as above.

michaeliwaikato commented on frontend-react/src/components/annotate.js:
```diff
@@ -3,73 +3,164 @@ import ReactImageAnnotate from "react-image-annotate";
       }
+    }
+    //Cancel deletion
+    else{
+      return
+    }
```
Probably redundant?

michaeliwaikato commented on frontend-react/src/components/annotate.js:
```diff
@@ -3,73 +3,164 @@ import ReactImageAnnotate from "react-image-annotate";
+    }
+  }
+
+  incrementIndex(imagesIndex){
+    this.setState({
+      index: this.state.index = imagesIndex.length
+    });
+  }
```
Could cause problems with asynchronous nature as above.

michaeliwaikato commented on frontend-react/src/components/annotate.js:
```diff
@@ -3,73 +3,164 @@ import ReactImageAnnotate from "react-image-annotate";
+  }
+
+  //Each time the annotator is instantiated, the current index is saved
+  updateIndex(index){
+    this.setState({
+      index: this.state.index=index.selectedImage
+    });
+  }
```
Same again.

michaeliwaikato commented on frontend-react/src/components/annotate.js:
```diff
@@ -3,73 +3,164 @@ import ReactImageAnnotate from "react-image-annotate";
+    });
+  }
+
+  saveAnnotation(annotateData){
+    //Save current images and annotations to state data
+    this.setState({
+      data: annotateData.images,
+      saved: true
+    });
+
+
+  }
+
```
And again.

michaeliwaikato commented on frontend-react/src/components/annotate.js:
```diff
@@ -3,73 +3,164 @@ import ReactImageAnnotate from "react-image-annotate";
+  }
+
+  //Open file input dialog
+  clickFileInput(){
+
+    var element = document.querySelector('#root > div > form > label')
+    const mouseClickEvents = ['mousedown', 'click', 'mouseup'];
+    mouseClickEvents.forEach(mouseEventType =>
+      element.dispatchEvent(
+        new MouseEvent(mouseEventType, {
+            view: window,
+            bubbles: true,
+            cancelable: true,
+            buttons: 1
+        })
+      ),
+    );
+  }
```
This can fairly easily be shortened and simplified to something like:

`document.getElementById('file').click()`

michaeliwaikato commented on frontend-react/src/components/annotate.js:
```diff
@@ -3,73 +3,164 @@ import ReactImageAnnotate from "react-image-annotate";
+    //If annotator has images uploaded
+    if(uploaded && data.length > 0){
+      return (
+      <div className="border">
```
```suggestion
      <div id="border">
```

michaeliwaikato commented on frontend-react/src/components/annotate.js:
```diff
@@ -3,73 +3,164 @@ import ReactImageAnnotate from "react-image-annotate";
+      return (
+      <div className="border">
+
+        <form className="selectForm">
```
```suggestion
        <form id="selectForm">
```

Are these forms actually needed anyway?

michaeliwaikato commented on frontend-react/src/components/annotate.js:
```diff
@@ -3,73 +3,164 @@ import ReactImageAnnotate from "react-image-annotate";
+      <div className="border">
+
+        <form className="selectForm">
+            <input type="file" id="file" multiple={true} accept="image/*" className="fileInputButton" onChange={this.fileSelectedHandler} />
```
```suggestion
            <input type="file" id="file" multiple={true} accept="image/*" onChange={this.fileSelectedHandler} />
```

michaeliwaikato commented on frontend-react/src/components/annotate.js:
```diff
@@ -3,73 +3,164 @@ import ReactImageAnnotate from "react-image-annotate";
+
+        <form className="selectForm">
+            <input type="file" id="file" multiple={true} accept="image/*" className="fileInputButton" onChange={this.fileSelectedHandler} />
+            <label htmlFor="file" className="fileLabel">
+              Add Images
+            </label>
```
```suggestion
```

michaeliwaikato commented on frontend-react/src/components/annotate.js:
```diff
@@ -3,73 +3,164 @@ import ReactImageAnnotate from "react-image-annotate";
+            </label>
+        </form>
+
+        <div className="imageDisplay">
```
```suggestion
        <div id="imageDisplay">
```

michaeliwaikato commented on frontend-react/src/components/annotate.js:
```diff
@@ -3,73 +3,164 @@ import ReactImageAnnotate from "react-image-annotate";
+        <div className="imageDisplay">
+
+          <ReactImageAnnotate
+            images= {data}
+            key = {data}
+            updateIndex = {output => this.updateIndex(output.selectedImage)}
+            regionClsList={[]}
+            enabledTools="create-polygon, create-box"
+            deleteImage={output => this.delete(output)}
+            addImages={output => {
+              this.saveAnnotation(output);
+              this.clickFileInput();
+            }}
+            selectedImage = {index}
+            onExit={output => console.log(JSON.stringify(output.images))}
```
Spacing etc should be consistent to make it easier to read/understand.

michaeliwaikato commented on frontend-react/src/components/annotate.js:
```diff
@@ -3,73 +3,164 @@ import ReactImageAnnotate from "react-image-annotate";
+
+        </div>
+      </div>
+    );}
```
```suggestion
    );
  }
```

michaeliwaikato commented on frontend-react/src/components/annotate.js:
```diff
@@ -3,73 +3,164 @@ import ReactImageAnnotate from "react-image-annotate";
+      </div>
+    );}
+    //If no images have been uploaded show "upload images" button
+    else{
+      return (
+        <div className="border">
+          <div className="imageDisplay">
+
+            <form className="selectForm" id="formUpload">
+                <input type="file" id="file" multiple={true} accept="image/*" className="fileInputButton" onChange={this.fileSelectedHandler} />
+                <label htmlFor="file" className="fileLabel">
+                    Upload Images
+                </label>
+            </form>
+
+          </div>
+        </div>
```
Same as above in terms of the ids etc.

An idea for this is that this is removed and a base sample image is added to the annotator showing some instructions on how to use. Then the "add image" button is clicked which clears the sample image and shows the users images.

michaeliwaikato commented on frontend-react/src/index.js:
```diff
@@ -5,9 +5,7 @@ import Navigation from './navigation';
 import * as serviceWorker from './serviceWorker';
 
 ReactDOM.render(
-  <React.StrictMode>
-    <Navigation />
-  </React.StrictMode>,
```
If you can indicate exactly why this was removed then that would be good.

michaeliwaikato commented on frontend-react/src/style/annotate.css:
```diff
@@ -25,6 +25,7 @@
   border-radius: 5px;
   font-size:20px;
   font-weight: bold;
+
```
```suggestion
```

michaeliwaikato commented on frontend-react/src/style/annotate.css:
```diff
@@ -25,6 +25,7 @@
+
 }
 
 .fileInputButton:focus + .fileLabel {
```
I can't add comments for the parts above, but these should be ids and not classes.

michaeliwaikato commented on frontend-react/src/style/annotate.css:
```diff
@@ -33,29 +34,30 @@
   height: 100%;
-  max-height: 90vh;
-  /* max-width: 90vh; */ 
+ /* max-height: 90vh;
+  /* max-width: 90vh; */
```
What are these for?

michaeliwaikato commented on frontend-react/src/style/annotate.css:
```diff
@@ -33,29 +34,30 @@
 }
 
 .selectForm {
-  margin-top: 20px;
+ /*  margin-top: 20px;*/
+  visibility:hidden;
```
Redundant.

michaeliwaikato commented on frontend-react/src/style/annotate.css:
```diff
@@ -33,29 +34,30 @@
 
 #formUpload {
   margin-top: 20%;
+  visibility: visible;
```
Redundant.

michaeliwaikato commented on frontend-react/src/style/annotate.css:
```diff
@@ -33,29 +34,30 @@
 }
 
 #formUpload {
   margin-top: 20%;
```
Seems a little hacky.

michaeliwaikato commented on frontend-react/src/style/annotate.css:
```diff
@@ -33,29 +34,30 @@
 }
 
 .imageDisplay {
```
Same as above.

Commit:
Tidied up annotation page

This commit tidies up the annotation page and fixes some of the functionality.

Commit:
Fixed annotation saving and updated dependencies

This commit re-enables existing annotations to be preserved when new images
are added. It also fixes the project dependencies so that it depends on
the custom annotation library.
